### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by juani-lastra**
Sprint 4 Development:

- Improvements 

- Search assets from DB, internet connection check 

- Barcode validation from input 

- get locations from firebase 

- Loaders implementation

### CODE REVIEW

File: app/src/main/java/com/mobiletandil/stockapp/activity/AssetFormActivity.kt [hunk 1]
```diff
@@ -8,161 +8,240 @@ import androidx.appcompat.app.AppCompatActivity
...
         binding.buttonSaveForm.text = getString(R.string.label_update_button)
 
         binding.editTextBarcode.setText(assetInformation.barcode)
         binding.editAssetName.setText(assetInformation.assetName)
         binding.editUserMail.setText(assetInformation.userMail)
-        binding.locationAutocompleteTextView.setText(assetInformation.location)
+        enableFieldsButBarcode()
+    }
+
+    override fun onBackPressed() {
```
Comment (SartiMau) [inline app/src/main/java/com/mobiletandil/stockapp/activity/AssetFormActivity.kt:243]: this is not correct, please remove this method

File: app/src/main/java/com/mobiletandil/stockapp/utils/GeneralUtils.kt [hunk 2]
```diff
@@ -1,17 +1,21 @@
...
+import android.view.inputmethod.InputMethodManager
 import android.widget.Toast
 import androidx.core.content.ContextCompat
+import androidx.fragment.app.Fragment
 
 object GeneralUtils {
     val REQUIRED_PERMISSIONS = arrayOf(Manifest.permission.CAMERA)
     const val REQUEST_CODE_PERMISSIONS = 10
-    private const val GLOBANT_MAIL = "@globant.com"
+    private const val ARROBA = "@"
```
Comment (SartiMau) [inline app/src/main/java/com/mobiletandil/stockapp/utils/GeneralUtils.kt:18]: please change the name to "AT"

File: app/src/main/java/com/mobiletandil/stockapp/viewModel/AssetFormViewModel.kt [hunk 3]
```diff
@@ -82,11 +92,34 @@ class AssetFormViewModel(private val saveAssetUseCase: SaveAssetUseCase) : ViewM
...
-    fun initUI(automaticBarcode: String, comeFromSearch: Boolean, assetInformation: Asset?) {
-        mutableAssetForm.value = when {
-            comeFromSearch -> Event(AssetFormData(state = AssetFormState.INIT_UI_FROM_SEARCH, asset = assetInformation))
-            automaticBarcode.isNotEmpty() -> Event(AssetFormData(state = AssetFormState.INIT_UI_WITH_BARCODE, barcode = automaticBarcode))
-            else -> Event(AssetFormData(state = AssetFormState.INIT_UI_WITHOUT_BARCODE))
+    fun initUI(automaticBarcode: String, comeFromSearch: Boolean, assetInformation: Asset?) = viewModelScope.launch {
+        withContext(Dispatchers.IO) { getLocationsUseCase() }.let {
+            var locations: List<Location> = emptyList()
+            when (it) {
+                is CoroutineResult.Success<List<Location>> -> {
```
Comment (SartiMau) [inline app/src/main/java/com/mobiletandil/stockapp/viewModel/AssetFormViewModel.kt:99]: What happens if it comes with CoroutineResult.Failure?

File: app/src/main/res/values/styles.xml [hunk 4]
```diff
@@ -51,6 +52,7 @@
         <item name="android:drawableLeft">@drawable/outline_manage_search_24</item>
         <item name="android:text">@string/title_search_button_main</item>
         <item name="android:layout_marginTop">@dimen/margin_top_search_button_main</item>
+        <item name="android:layout_marginBottom">250dp</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml:55]: do not hardcode pls, use dimens

File: app/src/main/res/values/styles.xml [hunk 5]
```diff
@@ -116,21 +118,21 @@
...
-        <item name="android:hint">@string/add_form_barcode_hint_text</item>
+        <item name="android:hint">@string/asset_form_barcode_hint_text</item>
         <item name="android:drawableRight">@drawable/ic_baseline_search_24</item>
-        <item name="android:maxLines">1</item>
+        <item name="android:maxLines">@integer/max_lines</item>
     </style>
 
     <style name="EditTextAssetNameStyle" parent="EditTextStyle">
-        <item name="android:hint">@string/add_form_asset_name_hint_text</item>
+        <item name="android:hint">@string/asset_form_asset_name_hint_text</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml:127]: Do not put hints in styles as they are specific for each field. Add the hint in the view and remove this style

File: app/src/main/res/values/styles.xml [hunk 6]
```diff
@@ -116,21 +118,21 @@
...
     </style>
 
     <style name="EditTextUserMailStyle" parent="EditTextStyle">
-        <item name="android:hint">@string/add_form_user_mail_hint_text</item>
+        <item name="android:hint">@string/asset_form_user_mail_hint_text</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml:131]: Do not put hints in styles as they are specific for each field. Add the hint in the view and remove this style

File: app/src/main/res/values/styles.xml [hunk 7]
```diff
@@ -104,7 +106,7 @@
     <style name="LocationLabel">
         <item name="android:textSize">@dimen/size_text_button_main</item>
         <item name="android:fontFamily">@font/heebo</item>
-        <item name="android:text">@string/add_form_location_label</item>
+        <item name="android:text">@string/asset_form_location_label</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml:109]: Do not put text in styles as they are specific for each field.

File: app/src/main/res/values/styles.xml [hunk 8]
```diff
@@ -85,15 +87,15 @@
...
+        <item name="android:text">@string/asset_form_title</item>
         <item name="android:textColor">@color/main_title_with_dirty</item>
         <item name="android:layout_width">match_parent</item>
         <item name="android:layout_height">wrap_content</item>
         <item name="android:layout_marginTop">@dimen/margin_textview_main</item>
     </style>
 
     <style name="DescriptionAddFormActivityStyle">
-        <item name="android:text">@string/add_form_description</item>
+        <item name="android:text">@string/asset_form_description</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml:98]: Do not put text in styles as they are specific for each field.

File: app/src/main/res/values/styles.xml [hunk 9]
```diff
@@ -85,15 +87,15 @@
         <item name="android:shadowDx">@integer/size_shadow_title_main</item>
         <item name="android:shadowDy">@integer/size_shadow_title_main</item>
         <item name="android:shadowRadius">@integer/size_shadow_radius_title_main</item>
-        <item name="android:text">@string/add_form_title</item>
+        <item name="android:text">@string/asset_form_title</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml:90]: Do not put text in styles as they are specific for each field.

File: app/src/main/res/values/styles.xml [hunk 10]
```diff
@@ -60,13 +62,13 @@
...
     <style name="ButtonAddFormSave" parent="ButtonMain">
-        <item name="android:text">@string/add_form_save_label</item>
+        <item name="android:text">@string/asset_form_save_label</item>
         <item name="android:layout_marginTop">@dimen/margin_textview_main</item>
         <item name="android:layout_marginBottom">@dimen/padding_dialog</item>
     </style>
 
     <style name="ButtonDialog" parent="ButtonMain">
-        <item name="android:text">@string/add_form_save_label</item>
+        <item name="android:text">@string/asset_form_save_label</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml:71]: Do not put text in styles as they are specific for each field.

File: app/src/main/res/values/styles.xml [hunk 11]
```diff
@@ -60,13 +62,13 @@
     </style>
 
     <style name="ButtonAddFormSave" parent="ButtonMain">
-        <item name="android:text">@string/add_form_save_label</item>
+        <item name="android:text">@string/asset_form_save_label</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml:65]: Do not put text in styles as they are specific for each field.

File: app/src/main/res/values/styles.xml [hunk 12]
```diff
@@ -11,6 +11,7 @@
         <item name="android:shadowRadius">@integer/size_shadow_radius_title_main</item>
         <item name="android:text">@string/title_main</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml:12]: Do not put text in styles as they are specific for each field.

File: app/src/main/res/values/styles.xml [hunk 13]
```diff
@@ -116,21 +118,21 @@
     </style>
 
     <style name="EditTextBarcodeStyle" parent="EditTextStyle">
-        <item name="android:hint">@string/add_form_barcode_hint_text</item>
+        <item name="android:hint">@string/asset_form_barcode_hint_text</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml:121]: Do not put hint in styles as they are specific for each field.

File: app/src/main/res/values/styles.xml [hunk 14]
```diff
@@ -116,21 +118,21 @@
...
     </style>
 
     <style name="EditTextUserMailStyle" parent="EditTextStyle">
-        <item name="android:hint">@string/add_form_user_mail_hint_text</item>
+        <item name="android:hint">@string/asset_form_user_mail_hint_text</item>
     </style>
 
     <style name="EditTextLocationStyle" parent="Widget.MaterialComponents.TextInputLayout.FilledBox.ExposedDropdownMenu">
-        <item name="android:hint">@string/add_form_location_hint_text</item>
+        <item name="android:hint">@string/asset_form_location_hint_text</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml:135]: Do not put hint in styles as they are specific for each field.

**Review by SartiMau — COMMENTED** [review]
app/src/test/java/com/mobiletandil/stockapp/viewmodel/AddFormViewModelTest.kt is empty, please remove it

File: app/src/main/java/com/mobiletandil/stockapp/activity/AssetFormActivity.kt (hunk 1 above)
Comment (Georgina-Fernandez) [reply app/src/main/java/com/mobiletandil/stockapp/activity/AssetFormActivity.kt:243]: Already in ticket: Improvements from merge to main

File: app/src/main/java/com/mobiletandil/stockapp/utils/GeneralUtils.kt (hunk 2 above)
Comment (Georgina-Fernandez) [reply app/src/main/java/com/mobiletandil/stockapp/utils/GeneralUtils.kt:18]: Already in ticket: Improvements from merge to main

File: app/src/main/java/com/mobiletandil/stockapp/viewModel/AssetFormViewModel.kt (hunk 3 above)
Comment (Georgina-Fernandez) [reply app/src/main/java/com/mobiletandil/stockapp/viewModel/AssetFormViewModel.kt:99]: Already in ticket: Improvements from merge to main

File: app/src/main/res/values/styles.xml (hunk 4 above)
Comment (Georgina-Fernandez) [reply app/src/main/res/values/styles.xml:55]: Already in ticket: Improvements from merge to main

File: app/src/main/res/values/styles.xml (hunk 5 above)
Comment (Georgina-Fernandez) [reply app/src/main/res/values/styles.xml:127]: Already in ticket: Improvements from merge to main

File: app/src/main/res/values/styles.xml (hunk 11 above)
Comment (Georgina-Fernandez) [reply app/src/main/res/values/styles.xml:65]: Already in ticket: Improvements from merge to main

File: app/src/main/res/values/styles.xml [hunk 15]
```diff
@@ -51,6 +52,7 @@
         <item name="android:drawableLeft">@drawable/outline_manage_search_24</item>
         <item name="android:text">@string/title_search_button_main</item>
```
[the remark this answers is not in the document]
Comment (Georgina-Fernandez) [reply app/src/main/res/values/styles.xml:53]: Already in ticket: Improvements from merge to main

File: app/src/main/res/values/styles.xml (hunk 12 above)
Comment (Georgina-Fernandez) [reply app/src/main/res/values/styles.xml:12]: Already in ticket: Improvements from merge to main

File: app/src/main/res/values/styles.xml (hunk 13 above)
Comment (Georgina-Fernandez) [reply app/src/main/res/values/styles.xml:121]: Already in ticket: Improvements from merge to main

**Comment by Georgina-Fernandez** [comment]
Link to Trello ticket with improvement detail: https://trello.com/c/ETJA4Ayl

**Review by SartiMau — APPROVED** [review]

File: app/src/main/java/com/mobiletandil/stockapp/fragment/SearchPickerFragment.kt [hunk 16]
```diff
@@ -62,32 +78,78 @@ class SearchPickerFragment : DialogFragment() {
...
                     )
                 }
             }
             SearchPickerStatus.SEARCH -> {
-                startActivity(activity?.let { AssetInformationListActivity.newIntent(it, eventData.filter, eventData.value) })
+                startActivity(
+                    activity?.let {
+                        eventData.filter?.let { it1 ->
+                            eventData.value?.let { it2 ->
+                                AssetInformationListActivity.newIntent(it, it1, it2)
```
Comment (mariano98montero) [inline app/src/main/java/com/mobiletandil/stockapp/fragment/SearchPickerFragment.kt:101]: please use representative names for this let{}

**Review by mauroLonghi — APPROVED** [review]

File: app/src/main/java/com/mobiletandil/stockapp/fragment/SearchPickerFragment.kt (hunk 16 above)
Comment (Georgina-Fernandez) [reply app/src/main/java/com/mobiletandil/stockapp/fragment/SearchPickerFragment.kt:101]: Already in ticket: Improvements from merge to main

**Pull request merged**